scanpy/plotting/_anndata.py

```diff
@@ -3507,3 +2696,1868 @@ def _check_var_names_type(var_names, var_group_labels, var_group_positions):
+        yticklabels
+            Because the plots are on top of each other the yticks labels tend to
+            overlap and are not plotted. Set to true to view the labels.
+        ylim
+            minimum and maximum values for the y-axis. If set. All rows will have
+            the same y-axis range. Example: ylim=(0, 5)
+
+        Returns
+        -------
+        StackedViolin
+
+        Examples
+        -------
+        >>> adata = sc.datasets.pbmc68k_reduced()
+        >>> markers = ['C1QA', 'PSAP', 'CD79A', 'CD79B', 'CST3', 'LYZ']
+
+        Change color map and turn off edges
+        >>> sc.pl.MatrixPlot(adata, markers, groupby='bulk_labels')\
+        ...               .style(row_palette='Blues', linewisth=0).show()
+
+        """
+
+        self.row_palette = row_palette
+        self.stripplot = stripplot
+        self.jitter = jitter
+        self.jitter_size = jitter_size
+        self.yticklabels = yticklabels
+        self.ylim = ylim
+
+        self.kwds['linewidth'] = linewidth
+        self.kwds['scale'] = scale
+
+        # set by default the violin plot cut=0 to limit the extend
+        # of the violin plot as this produces better plots that wont extend
+        # to negative values for example. From seaborn.violin documentation:
+        #
+        # cut: Distance, in units of bandwidth size, to extend the density past
+        # the extreme datapoints. Set to 0 to limit the violin range within
+        # the range of the observed data (i.e., to have the same effect as
+        # trim=True in ggplot.
+        self.kwds.setdefault('cut', 0)
+        self.kwds.setdefault('inner')
+        return self
+
+    def _mainplot(self, ax):
+
+        # work on a copy of the dataframes. This is to avoid changes
+        # on the original data frames after repetitive calls to the
+        # DotPlot object, for example once with swap_axes and other without
+        _matrix = self.obs_tidy.copy()
+
+        global count
+        count = 0
+
+        def rename_cols_to_int(value):
+            global count
+            count += 1
+            return count
+
+        if self.var_names_idx_order is not None:
+            _matrix = _matrix.iloc[:, self.var_names_idx_order]
+
+        if self.categories_order is not None:
+            _matrix.index = _matrix.index.reorder_categories(self.categories_order, ordered=True)
+
+        row_palette = self.kwds.get('color', self.row_palette)
+        if 'color' in self.kwds:
+            del self.kwds['color']
+
+        # All columns should have a unique name, yet, frequently
+        # gene names are repeated in self.var_names.  otherwise the
+        # pd.melt object that is passed to seaborn will merge non-unique columns.
+        # Here, I simply rename the columns using a count from 1..n using the
+        # mapping function `rename_cols_to_int` to solve the problem.
+        _matrix.rename(rename_cols_to_int, axis='columns', inplace=True)
```

Comment:
Defining a global variable is not a good solution to this problem.

Could just do `df.rename(columns=dict(zip(df.columns, range(len(df.columns)))))`?

Reply:
I remember I had some issue in the past (this code is quite old btw)  which I could not resolve cleanly and had to resort to these weird function. I will try your method.